sql/rpl_binlog_sender.cc

```diff
@@ -654,29 +655,30 @@ int Binlog_sender::send_events(File_reader &reader, my_off_t end_pos) {
         m_packet.copy(tmp);
         m_packet.length(tmp.length());
       }
+      needs_send_packet = true;
```

Comment:
what about if (send_heartbeat_event(log_pos)) at line 632? do we need to set needs_send_packet=true for that case?

Reply:
Did you mean ignore calling send_packet()? Becuase after your change, we go back to pre 8.0.31 behaviour and the before/after_send_hook()s are executed independently of send_packet(). Isnt it?

Reply:
I meant avoiding the call to before_send and after_send hooks for `send_heartbeat_event()`. After this patch we will be calling those hooks only when `send_packet()` is called.

Reply:
Hi! I am sorry for the confusion and please ignore my earlier comments. 

We must call before_send and after_send hooks for heartbeat events to fix this bug.

So, I have further simplified the patch so that the behavior before 8.0.31 is preserved, and upstream bug fix is also honored.

New logic is
```
if (events need to be skipped) {
  before_send_hook()
  send_heartbeat()
  after_send_hook()
} else {
  Create Observe_transmission_guard
  before_send_hook()
  send_packet()
  after_send_hook()
}
```

Reply:
Thanks!